storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java

```diff
@@ -0,0 +1,553 @@
+     * and page #10 from disk (page fault), which can take up to more than a second. In our test, this can cause the
+     * at-least-once produce latency to jump to about 1 second from a few ms.
+     *
+     * Here, we use a more cache-friendly lookup algorithm:
+     * if (target > indexEntry[end - N]) // if the target is in the last N entries of the index
+     *    binarySearch(end - N, end)
+     * else
+     *    binarySearch(begin, end - N)
+     *
+     * If possible, we only look up in the last N entries of the index. By choosing a proper constant N, all the in-sync
+     * lookups should go to the 1st branch. We call the last N entries the "warm" section. As we frequently look up in this
+     * relatively small section, the pages containing this section are more likely to be in the page cache.
+     *
+     * We set N (_warmEntries) to 8192, because
+     * 1. This number is small enough to guarantee all the pages of the "warm" section is touched in every warm-section
+     *    lookup. So that, the entire warm section is really "warm".
+     *    When doing warm-section lookup, following 3 entries are always touched: indexEntry(end), indexEntry(end-N),
+     *    and indexEntry((end*2 -N)/2). If page size >= 4096, all the warm-section pages (3 or fewer) are touched, when we
+     *    touch those 3 entries. As of 2018, 4096 is the smallest page size for all the processors (x86-32, x86-64, MIPS,
+     *    SPARC, Power, ARM etc.).
+     * 2. This number is large enough to guarantee most of the in-sync lookups are in the warm-section. With default Kafka
+     *    settings, 8KB index corresponds to about 4MB (offset index) or 2.7MB (time index) log messages.
+     *
+     *  We can't set make N (_warmEntries) to be larger than 8192, as there is no simple way to guarantee all the "warm"
+     *  section pages are really warm (touched in every lookup) on a typical 4KB-page host.
+     *
+     * In there future, we may use a backend thread to periodically touch the entire warm section. So that, we can
+     * 1) support larger warm section
+     * 2) make sure the warm section of low QPS topic-partitions are really warm.
+     */
+    protected final int warmEntries() {
+        return 8192 / entrySize();
+    }
+
+    protected void safeForceUnmap() {
+        if (mmap != null) {
+            try {
+                forceUnmap();
+            } catch (Throwable t) {
+                log.error("Error unmapping index {}", file, t);
+            }
+        }
+    }
+
+    /**
+     * Forcefully free the buffer's mmap.
+     */
+    // Visible for testing, we can make this protected once OffsetIndexTest is in the same package as this class
+    public void forceUnmap() throws IOException {
+        try {
+            ByteBufferUnmapper.unmap(file.getAbsolutePath(), mmap);
+        } finally {
+            mmap = null;
+        }
+    }
+
+    // GuardedBy `lock`
```

Comment:
Should we change the comment to the following?

The caller is expected to hold `lock` when calling this method.

Reply:
I wanted to use the @GuardedBy annotation, but there are some licensing issues with the jcip jar (LGPL). I'll go with your suggested comment for now.